File: ruby/sum_of_left_leaves.rb
def sum_of_left_leaves(root)
  sum = 0
  return sum unless root
  stack = []
  stack.push([root, false])

  while(stack.any?) do
    node, is_left = stack.pop
    if node.left || node.right
      stack.push([node.left, true]) if node.left
      stack.push([node.right, false]) if node.right
    else
      sum += node.val if is_left
    end
  end
  sum
end

File: java/SumOfLeftLeaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int sumOfLeftLeaves(TreeNode root) {
        return sum(root, 0);
    }

    public int sum(TreeNode node, int sum) {
        if(node == null) return sum;
        if(node.left != null) {
            if(node.left.left == null && node.left.right == null) {
                sum += node.left.val;
            }
            sum = sum(node.left, sum);
        }

        sum = sum(node.right, sum);

        return sum;
    }

    public class TreeNode {
      int val;
      TreeNode left;
      TreeNode right;
      TreeNode(int x) { val = x; }
    }
}
